core/server/master/src/main/java/alluxio/master/file/meta/InodeTreePersistentState.java

```diff
@@ -280,31 +291,34 @@ public void applyAndJournal(Supplier<JournalContext> context, MutableInode<?> in
 
   private void applyDelete(DeleteFileEntry entry) {
     long id = entry.getId();
-
     Inode inode = mInodeStore.get(id).get();
 
-    mInodeStore.removeInodeAndParentEdge(inode);
-    updateLastModifiedAndChildCount(inode.getParentId(), entry.getOpTimeMs(), -1);
-    mPinnedInodeFileIds.remove(id);
-    mReplicationLimitedFileIds.remove(id);
-    mToBePersistedIds.remove(id);
-
     // The recursive option is only used by old versions.
     if (inode.isDirectory() && entry.getRecursive()) {
       Queue<InodeDirectory> dirsToDelete = new ArrayDeque<>();
       dirsToDelete.add(inode.asDirectory());
       while (!dirsToDelete.isEmpty()) {
         InodeDirectory dir = dirsToDelete.poll();
-        mInodeStore.remove(inode);
+        mInodeStore.removeInodeAndParentEdge(inode);
+        mInodeCounter.decrementAndGet();
```

Comment:
should the counter not be a part of the inode Store? There seems to be an `InodeStore.remove()` and `InodeStore.write()` which seems like natural places to update a count?

Reply:
The challenge is that the inode store doesn't know whether inodes are *actually* added/removed. removal is allowed on non-existent inodes and creation is allowed on already-existing inodes. This looseness in the API is necessary to support RocksDB as a type of inode store. To push the counting into the inode store, the inode store would need to assume that the caller will only call remove() on existing inodes and create() on non-existing inodes